### ISSUE TITLE
Simplified typing

### ISSUE THREAD

**Issue by christianalfoni** [cerebral/overmind#481]
Dear Overminders!

I have one thing I really want to do and that is to clean up and simplify the typing of Overmind. It was a long process figuring out how to do it and with acquired knowledge through all this work and later projects I would like to do this:

```ts
import { IContext } from 'overmind'

const config = { ... }

export interface Context extends Content<{
  state: typeof config.state,
  actions: typeof config.actions,
  effects: typeof config.effects
}> {}
```

There will be no more **implicit typing**, this is what everybody will need to do. This is how you type the different things:

Actions and operators are typed by importing the `Context` and explicitly typing the first argument:
```ts
const myAction = (context: Context) => {}

const myActionWithPayload = (context: Context, payload: string) => {}

// Now we support generics
const myActionWithPayload = <T extends TEvent>(context: Context, payload: T) => {}


```

```ts

```

**Comment by christianalfoni** [comment]
As part of this change the operators are also extremely simplified, and best of all... you can just compose any action directly in as an operator.

```ts
const someAction = ({}: Context, value: string) => {}

const operator = pipe(
  someAction,
  wait(500)
)
```

It basically means operators finally are interoperable with plain actions. Which also means we do not need `mutate`, `map` and `run` anymore. You can just give it a plain action to do all those things. We rather have operators that goes beyond what a normal action can do.

This is such a huge improvement!

**Comment by suevalov** [comment]
Looks like a great improvement and it doesn't look like a lot of work to refactor our app.

**Comment by jmattheis** [comment]
Looks good, will there be errors if the signature of actions doesn't match (Context, Payload). f.ex. what would happen if an action similar to this is defined:
```
const myAction = (context: string) => {}
```

**Comment by christianalfoni** [comment]
@jmattheis Good question!

Yes, when the actions are attached to Overmind it expects the signature of: ´(context: IContext<any>)` , so you will get the error where you instantiate Overmind rather than where you define the action 😄

**Comment by christianalfoni** [comment]
So a summary of what is being moved into `next` now is:

**New typing, no more implicit typing**:

```ts
import { IContext } from 'overmind'

const config = { ... }

export interface Context extends IContext<{
  state: typeof config.state,
  actions: typeof config.actions,
  effects: typeof config.effects
}> {}
```

**Type actions**:
```ts
const myAction = (context: Context) => {}

const myActionWithPayload = (context: Context, payload: string) => {}
```

**Operator changes**:
- Now an operator is interoperable with actions, meaning you can just do:

```ts
const whatever = pipe(
  (context: Context, input: string) => input.toUpperCase()
)
```

- Due to interoperable actions, the `map`, `mutate` and `run` operators has been removed. Just use plain action signature

- `forEach` operators has also been removed as the typing never worked there (This might come back, have to research some more)

- `fork` has new signature:

```ts
pipe(
  // "type" is a property on the object on the pipe
  fork('type', {
    FOO: (context: Context, event) => {
      event // Is now typed as FooEvent
    },
    BAR: (context: context, event) => {
      event // Is now typed as BarEvent
    } 
  })
)
```

- `parallel` now actually returns an array to the pipe with the results of each operator

**React hooks**
- `useOvermind` is removed in favour of explicit hooks (more efficient and cleaner)
- `useState` hook now has an added signature

```ts
const ItemComponent = ({ id }) => {
  // The state you access within the callback is not tracked,
  // only the result and further access is tracked. Meaning this
  // component will only track the item, not changes to "items"
  const item = useAppState(state => state.items[id])
}
```

**Comment by christianalfoni** [comment]
Hi @ssijak , when you write this in the same file you will get this error. If you rather do:

```ts
export type Context = IContext<{
  state: typeof state,
  actions: typeof actions,
  effects: typeof effects
}>
```

Then it works within the same file. `config.*` only works across files. I guess this is some TypeScript bug, or just how it works. But typically you would not put it in the same file and there are no examples in the documentation. BUT! I totally see why you tried 😄

**Comment by ssijak** [comment]
Thanks for the quick answer! It works now, it does not complain. But I noticed some possible issues around actions. I am a relative typescript noob so maybe I'm talking nonsense but:

If I type action parameters as any like : 
```
const changeState = ({ state }: Context, value: any) => {
  state.value = value
}
```
than c.changeState bellow is inferred to `()=>void` type
```
export const useActions = createActionsHook<Context>();
const c = useActions()
c.changeState()
```

**Comment by christianalfoni** [comment]
Thanks for testing! Doing `any` will also match `void` I guess, which is how actions identify "no payload". I would strongly argue that `any` should not be used as a typed payload, but it is confusing for sure 😄 

`any` is typically used for unknown objects. Cause if it was like a string, number etc. you would actually type it. So doing `{}` instead will probably fix it, as it matches any object, but not `void`

But this is good for documentation for sure!

**Comment by ssijak** [comment]
Thanks. I encountered one more possible issue. I do not find `OnInitialize` exported from overmind in the next tag. Is it removed or?

**Comment by Michielnuyts** [comment]
@christianalfoni This is all coming in V28 right? Do you have a rough release date in mind?

**Comment by christianalfoni** [comment]
There has been a little bit lack of time lately (Blame the baby mostly 😉 ), but also at codesandbox.io we are hiring a few people and it has also taken up a lot of my time. 

But if you look at the issue list now I have marked some issues as `NEXT RELEASE`, those needs to be fixed. Good thing is that our codesandbox.io PR, https://github.com/codesandbox/codesandbox-client/pull/5517, is now running green and should be merged shortly. Which was also a blocker for the next release.

So yeah, I hope to get some time soon to get through the pending issue and get that release out 😄

**Comment by mattiasjosefsson** [comment]
@christianalfoni I wanted to try out these new typings with my current namespaced config where my state and actions are placed in a separate folder. However in order to access the new `Context` type it needs to be imported in the actions file which would cause a cyclic dependency, how would I go about avoiding this? I couldn't figure it out from your examples.

`overmind/index.ts`
```typescript
import { IContext } from 'overmind'
import { namespaced } from 'overmind/config'
import * as status from './status' //<---- cyclic dependency

export const config = namespaced({
  status,
})

export type Context = IContext<{
  state: typeof config.state
  actions: typeof config.actions
}>
```

`overmind/status/index.ts`
``` typescript
import { state } from './state'
import * as actions from './actions'

export { state, actions }
```

`overmind/status/actions.ts`
```typescript
import { debounce, pipe } from 'overmind'
import { Context } from '..' // <---- cyclic dependency

export const updateState = (context: Context, payload: string) => {}
```

Thanks for an awesome library!

**Comment by eirikhm** [comment]
@mattiasjosefsson Move config into a separate file :-)

**Comment by mattiasjosefsson** [comment]
@eirikhm Aaah of course! Thanks!

**Comment by mattiasjosefsson** [comment]
@eirikhm Finally got around to test it but it doesn't help to move config into a separate file because `actions` are still going to be referenced by `Context` in `overmind.ts`

**Comment by eirikhm** [comment]
@mattiasjosefsson  something like this:


`overmind/config.ts`

```
import { merge, namespaced } from "overmind/config"

import { state } from "./state"
import * as actions from "./actions"
import * as effects from "./effects"

import * as status from "./status"

export const config = merge(
  {
    state,
    actions,
    effects,
  },
  namespaced({
    status,
  })
)
```

`overmind/index.ts`

```
import {
    createActionsHook,
    createStateHook,
    createEffectsHook,
  
  } from "overmind-react";
  import { IContext } from "overmind";
  
  import { config } from "./config";
  
  export type Context = IContext<{
    state: typeof config.state,
    actions: typeof config.actions,
    effects: typeof config.effects,
  }>
  

  export type RootState = typeof config.state;
  
  export const useAppState = createStateHook<Context>()
  export const useActions = createActionsHook<Context>()
  export const useEffects = createEffectsHook<Context>()
```


`overmind/status/actions.ts`

```
import { Context } from "../index"

export const updateState = (context: Context, payload: string) => {}

```

**Comment by mattiasjosefsson** [comment]
@eirikhm Could it be just eslint that complains when it's not really an issue?

**Comment by avele** [comment]
@christianalfoni  So there is no way to simplify writing actions? Duplicating `export const` everywhere doesn't feel good. Ideally, for me, it would be object actions with local type definitions.

```ts
const actions = {
  example (context: Context, params: { a: string, b: boolean }) {}
}
```
Right now if I try to write this way I get the error:
```
'actions' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.ts(7022)
```

**Comment by Spring3** [comment]
@mattiasjosefsson no, it is not an issue with eslint. Is definitely a cyclic reference. I faced the same problem.

It doesn't matter if some portion of code gets moved into a separate file, because `Context` type can be defined only when the full config object is merged together, and full config object includes actions, which require `Context` to type their args.
